ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp

```diff
@@ -558,37 +558,51 @@ int get_color(int defcolor, bool advanced)
 
 string get_directory(string dname, string caption)
 {
-//NOTE: This uses the Windows Vista or later file chooser, which is different than the one used by GM8 and lower
-//because I could not find out which one it uses, since IFileDialog is used by both wxWidgets and QtFramework
-//and there doesn't appear to be a standard file picker for XP or lower in the Windows API except SHBrowseForFolder that is
-//used by Game Maker for get_directory_alt
-  IFileDialog* fileDialog;
-  CoCreateInstance(CLSID_FileOpenDialog, NULL, CLSCTX_INPROC_SERVER, IID_PPV_ARGS(&fileDialog));
+  //NOTE: This uses the Windows Vista or later file chooser, which is different than the one used by GM8 and lower
+  //because I could not find out which one it uses, since IFileDialog is used by both wxWidgets and QtFramework
+  //and there doesn't appear to be a standard file picker for XP or lower in the Windows API except SHBrowseForFolder that is
+  //used by Game Maker for get_directory_alt
+  IFileDialog *selectDirectory;
+  CoCreateInstance(CLSID_FileOpenDialog, NULL, CLSCTX_INPROC_SERVER, IID_PPV_ARGS(&selectDirectory));
 
   DWORD options;
-  fileDialog->GetOptions(&options);
-  options &= ~FOS_FILEMUSTEXIST;
-  options &= ~FOS_PATHMUSTEXIST;
-  fileDialog->SetOptions(options | FOS_PICKFOLDERS);
-  //TODO: Set default directory to dname
-  //fileDialog->SetDefaultFolder(std::wstring(dname.begin(), dname.end()).c_str());
-  fileDialog->SetTitle(std::wstring(caption.begin(), caption.end()).c_str());
-
-  fileDialog->Show(enigma::hWnd);
-
-  string res = "";
-  IShellItem *psi;
-
-  if (SUCCEEDED(fileDialog->GetResult(&psi))) {
-    LPWSTR wideres;
-    psi->GetDisplayName(SIGDN_DESKTOPABSOLUTEPARSING, &wideres);
-    psi->Release();
-
-    std::wstring wstr = wideres;
-    res = string(wstr.begin(), wstr.end());
+  selectDirectory->GetOptions(&options);
+  selectDirectory->SetOptions(options | FOS_PICKFOLDERS | FOS_NOCHANGEDIR | FOS_FORCEFILESYSTEM);
+
+  tstring tstr_dname = widen(dname);
+  LPWSTR szFilePath = (wchar_t *)tstr_dname.c_str();
+
+  IShellItem *pItem = nullptr;
+  HRESULT hr = ::SHCreateItemFromParsingName(szFilePath, nullptr, IID_PPV_ARGS(&pItem));
+
+  if (SUCCEEDED(hr)) {
+    LPWSTR szName = nullptr;
+    hr = pItem->GetDisplayName(SIGDN_NORMALDISPLAY, &szName);
+    if (SUCCEEDED(hr)) {
+      selectDirectory->SetFolder(pItem);
+      ::CoTaskMemFree(szName);
+    }
+    pItem->Release();
+  }
+
+  if (caption.empty()) selectDirectory->SetTitle(L"Select Directory");
```

Comment:
Does GM really do this if you don't set the caption? 🤕 Either way I don't like it, just curious though, and later on functions could be provided to allow the user to override the default caption. Something like `set_directory_default_caption` or something, I don't know.

Reply:
It actually does that on its own, but the default caption it has it "Select Folder". I changed it to "Select Directory" because that is the caption GM's function uses.

Reply:
Did you prefer it to say "Select Folder"? I can remove the excess code if you do.

Reply:
Oh, well, I guess that's fine. But I don't know if I would have bothered because of my comment above. If we made a fallback for Windows XP (not asking you to do this just to be clear) then we'd probably use the same API GM8.1 did for its outdated get directory dialogs. In that case, it would have that title. I would have just left the newer fallback use the one Windows picks, because, if theoretically YoYoGames ever decides to bring this back, that's likely what caption it would have. So just in general I wouldn't have bothered, but it's whatever.

Reply:
Well, the API dialog used for the original get_directory function is Delphi-specific. It only exists in Delphi applications, I know this from using Delphi and doing research on the matter. The best we could do is make an imitation dialog at best, like the one I made in the past for Dialog Module that I ended up replacing.

Reply:
Fair enough, let's just continue with the Windows standard one because that's likely what GMS would have if it ever brought back the function.

Reply:
Resolved: 61383af99eeee11165935b158a52158258144e6c